test/network/exception_test.cpp

```diff
@@ -0,0 +1,204 @@
+//===----------------------------------------------------------------------===//
+//
+//                         Peloton
+//
+// exception_test.cpp
+//
+// Identification: test/network/exception_test.cpp
+//
+// Copyright (c) 2016-17, Carnegie Mellon University Database Group
+//
+//===----------------------------------------------------------------------===//
+
+#include "common/harness.h"
+#include "gtest/gtest.h"
+#include "common/logger.h"
+#include "network/network_manager.h"
+#include "network/protocol_handler_factory.h"
+#include "util/string_util.h"
+#include <pqxx/pqxx> /* libpqxx is used to instantiate C++ client */
+#include <include/network/postgres_protocol_handler.h>
+#include <pqxx/except.hxx>
+#define NUM_THREADS 1
+
+namespace peloton {
+namespace test {
+
+//===--------------------------------------------------------------------===//
+// Exception Test
+//===--------------------------------------------------------------------===//
+
+class ExceptionTests : public PelotonTest {};
+
+static void *LaunchServer(peloton::network::NetworkManager network_manager,
+                          int port) {
+  try {
+    network_manager.SetPort(port);
+    network_manager.StartServer();
+  } catch (peloton::ConnectionException &exception) {
+    LOG_INFO("[LaunchServer] exception in thread");
+  }
+  return NULL;
+}
+
+/*
+ * To test with the queries with syntax error that will be caught by parser.
+ * The server will catch these errors in Networking layer and directly return ERROR response.
+ */
+void *ParserExceptionTest(int port) {
+  try {
+    // forcing the factory to generate psql protocol handler
+    pqxx::connection C(StringUtil::Format(
+        "host=127.0.0.1 port=%d user=postgres sslmode=disable application_name=psql", port));
+
+    peloton::network::NetworkConnection *conn =
+        peloton::network::NetworkManager::GetConnection(
+            peloton::network::NetworkManager::recent_connfd);
+
+    network::PostgresProtocolHandler *handler =
+        dynamic_cast<network::PostgresProtocolHandler*>(conn->protocol_handler_.get());
+    EXPECT_NE(handler, nullptr);
+
+    // If an exception occurs on one transaction, we can not use this transaction anymore
+    int exception_count = 0, total = 6;
+
+    // DROP query
+    pqxx::work txn1(C);
+    try {
+      txn1.exec("DROP TABEL IF EXISTS employee;");
+      txn1.commit();
+    } catch (const pqxx::pqxx_exception &e) {
+      const pqxx::sql_error *s = dynamic_cast<const pqxx::sql_error*>(&e.base());
+      if (s) {
+        LOG_INFO("Invalid Drop Query: %s", e.base().what());
+        exception_count += 1;
+      }
+    }
+
+    // CREATE query
+    pqxx::work txn2(C);
+    try {
+      // TODO(Yuchen): If the query is 'CREATE TABLE employee(id ITN)', postgres parser will generate a valid parse tree
```

Comment:
@haojin2 I am wondering, there may be a memory leak in the parser. Could you have a look it?
It happens when issuing query _CREATE TABLE employee(id ITN)_. It is invalid since the typename _ITN_ is wrong. The postgres parser returns a valid parse tree. But Peloton will catch the error when doing node transform. The memory block allocated for 'result' object in CreateTransform() is not freed. I see that ListTransform() will free the SQLStatementList. The memory allocated in CreateTransform() may still not be freed since the CreateStatement is not added into the SQLStatementList yet. I am thinking maybe we need some changes in the exception handling. Could you check that? Thanks

Reply:
Fixed in #979